Fix #17319: changes to fix the accessibility problems with preferences page

Patel-Muhammad opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #17319 
2. This PR does the following: 
- Added focus to the profile icon in top navbar.
- The menu options (which come while hovering the profile) were not having correct color contrast, I fixed it for better visibility.
- Added focus to the username.
- Added focus and fixed double selection of the buttons (Delete and Export) by adding tab index.
- Added focus to the footer anchor tags for visibility while selecting.
- Overall also added some tabindices to maintain the overall flow of the selection using tab.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

Before:

https://user-images.githubusercontent.com/96219910/221379312-80aff259-ddd8-4f72-89b2-005477d8c7c5.mp4

After:


https://user-images.githubusercontent.com/96219910/221379327-4da9e548-578c-4633-910b-8e945263f01d.mp4


<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

Patel-Muhammad committed:
Fixed issue: #17319

Patel-Muhammad committed:
Fixed lint errors

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad commented:
@vojtechjelinek can you please share your review for this PR?

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad commented:
@seanlip this PR is failing one test, can you please help me :
1. To fix the problem that is causing the test to fail.
2. Can any PR still be merged even though it is failing 1 or more tests?

seanlip commented:
No, PRs must pass CI to be mergeable.

For (1): 
- If the failure could be due to your changes, and you are having trouble fixing it, please contact your team leads @SanjaySajuJacob @Priyansh61 for help.
- If the failure is dev workflow related, please contact the dev workflow leads @gp201 @U8NWXD instead.

Thanks!

Patel-Muhammad commented:
@seanlip thanks for the clarification.

Patel-Muhammad commented:
@SanjaySajuJacob @Priyansh61 One test is failing for this PR, I tired checking any issue with the code that I changed, but I am not getting problem.

 Can you please check if the problem is there is my code change or it is dev workflow related.

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.html:
```diff
@@ -322,10 +323,8 @@ <h2 class="oppia-preferences-page-heading-subtext">
       <div class="col-lg-8 col-md-8 col-sm-8">
-        <a href="/delete-account" [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.DELETE_ACCOUNT.ROUTE">
-          <button type="button" class="btn btn-danger e2e-test-delete-account-button">
+        <a href="/delete-account"  class="btn btn-danger delete-btn e2e-test-delete-account-button focus-class" [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.DELETE_ACCOUNT.ROUTE">
```
```suggestion
        <a href="/delete-account"
        class="btn btn-danger delete-btn e2e-test-delete-account-button focus-class" 
        [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.DELETE_ACCOUNT.ROUTE">
```

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.css:
```diff
@@ -71,3 +71,11 @@
 .oppia-form-text {
   font-size: smaller;
 }
+.focus-class:focus-visible {
+  outline: 2px solid #0060df;
+  text-decoration: underline;
+}
+a.delete-btn, a.delete-btn:hover, a.delete-btn:visited,a.delete-btn:active {
+  color: #fff;
+  text-decoration: none;
+}
```
Please make sure these are prefixed with the preferences page class so that they are not applied in different places.

vojtechjelinek commented on core/templates/components/common-layout-directives/navigation-bars/side-navigation-bar.component.html:
```diff
@@ -274,7 +275,7 @@
       </div>
       <div [hidden]="!learnSubmenuIsShown" class="oppia-community-library-section">
-        <a href="/community-library" [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.LIBRARY_INDEX.ROUTE" class="e2e-mobile-test-library-link">
+        <a href="/community-library" [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.LIBRARY_INDEX.ROUTE" class="e2e-mobile-test-library-link" tabindex="-1">
```
```suggestion
        <a href="/community-library" 
        [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.LIBRARY_INDEX.ROUTE" 
        class="e2e-mobile-test-library-link" 
        tabindex="-1">
```

vojtechjelinek reviewed (changes requested):
Thanks! I have a few comments. Also in your video for the profile picture there are basically two "stops" one on the whole image the other specifically on the pencil icon, is that needed? I think the behaviour of both of those is the same.

Patel-Muhammad committed:
Update core/templates/pages/preferences-page/preferences-page.component.html

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Patel-Muhammad committed:
Update core/templates/components/common-layout-directives/navigation-bars/side-navigation-bar.component.html

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Removed the tab index from profile img, replaced button- Export accounto anchor tag

Patel-Muhammad committed:
Changed color for the export button

Patel-Muhammad committed:
Fixed lint issues

Patel-Muhammad committed:
Fixed lint-2

Patel-Muhammad commented:
@vojtechjelinek I have addressed your reviews, and committed the  necessary changes. I also changed one more thing as a better practice: 
- I found on this [html standards resource](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-a-element) that it is a bad practice to keep a button inside an anchor tag, so I removed the button and now we just have the anchor tag which does the same thing.

Video proof of the changes:

https://user-images.githubusercontent.com/96219910/222530767-fe96e02a-8fdd-467e-b8d6-b2b34b6c4953.mp4

vojtechjelinek commented on core/templates/pages/preferences-page/preferences-page.component.css:
```diff
@@ -71,3 +71,11 @@
+  outline: 2px solid #0060df;
+  text-decoration: underline;
+}
+a.delete-btn, a.delete-btn:hover, a.delete-btn:visited,a.delete-btn:active,a.export-btn, a.export-btn:hover, a.export-btn:visited,a.export-btn:active {
```
```suggestion
a.delete-btn, 
a.delete-btn:hover, 
a.delete-btn:visited,
a.delete-btn:active,
a.export-btn, 
a.export-btn:hover, 
a.export-btn:visited,
a.export-btn:active {
```

vojtechjelinek reviewed (changes requested):
Thanks! Just one nit. Also just to check is the whole profile picture still clickable?

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Update core/templates/pages/preferences-page/preferences-page.component.css

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

Patel-Muhammad commented:
@vojtechjelinek I just want to confirm, are you asking : that is the profile picture (not the pen) is clickable using mouse click?

vojtechjelinek commented:
> @vojtechjelinek I just want to confirm, are you asking : that is the profile picture (not the pen) is clickable using mouse click?

Yes

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad commented:
@vojtechjelinek here is the video showing that the whole profile picture still clickable:


https://user-images.githubusercontent.com/96219910/223221068-856762c7-1564-4d4b-9bb1-21a2f1107ddd.mp4

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

vojtechjelinek reviewed (approved):
LGTM!

Patel-Muhammad commented:
Hey @seanlip and @Priyansh61 ,  this PR is failing the lighthouse test and here is the log:

<img width="437" alt="Screenshot 2023-03-09 130547" src="https://user-images.githubusercontent.com/96219910/223952454-aee2af60-02ee-4cf4-9442-6db596b59753.png">

The problem:
I have not done any changes in any files related to `/topics-and-skills-dashboard`  page or `/skill_editor` page, still these pages are failing the lighthouse test, can you please help me solve the problem failing test. 
Thanks.

heyimShivam commented on core/templates/components/common-layout-directives/navigation-bars/top-navigation-bar.component.css:
```diff
@@ -407,7 +407,7 @@
   color: #009688;
 }
 .oppia-top-navigation-bar .oppia-navbar-dropdown .nav-link:hover {
-  color: #888;
+  background-color: #d7d7d7;
```
?

Patel-Muhammad replied:
@heyimShivam as you can see the before and after video of this pull request https://github.com/oppia/oppia/pull/17490#issue-1599841419
here on clicking the enter on the profile icon of the navbar the dropdown is opening, but we are not able to see what is selected clearly as there is no clear color indication, so by this I have changed the background so that it is visible what is selected.

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad commented:
@SanjaySajuJacob can you please review this PR? Thanks.

SanjaySajuJacob reviewed (commented):
Hey @Patel-Muhammad , would you take a look at why the accessibility tests are failing. Thanks!

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
focus-visible --> focus

Patel-Muhammad committed:
Revert "focus-visible --> focus"

This reverts commit a654d5fa823015a8d22c0d4d875b910482f55321.

Patel-Muhammad committed:
added aria-label

Patel-Muhammad committed:
Merge branch 'develop' into pre-patch

Patel-Muhammad committed:
Merge branch 'develop' of https://github.com/oppia/oppia into pre-patch

Patel-Muhammad committed:
Fixed lighthouse score

Patel-Muhammad committed:
Merge branch 'pre-patch' of https://github.com/Patel-Muhammad/oppia into pre-patch

Patel-Muhammad committed:
Fixed lint

Patel-Muhammad committed:
Merge branch 'develop' of https://github.com/oppia/oppia into pre-patch

Patel-Muhammad committed:
Fixed accessibility

SubhashThenua committed:
Merge branch 'develop' into pre-patch
